jobs/job_test_utils.py

```diff
@@ -0,0 +1,149 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities for running Apache Beam tests."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+
+from core.tests import test_utils
+from jobs import base_jobs
+from jobs import job_options
+from jobs.io import stub_io
+
+from apache_beam import runners
+from apache_beam.testing import test_pipeline
+from apache_beam.testing import util as beam_testing_util
+
+
+class BeamTestBase(test_utils.TestBase):
+    """Base Apache Beam test class with helpful constants and assertions."""
+
+    NOW = datetime.datetime.utcnow()
+    YEAR_AGO = NOW - datetime.timedelta(weeks=52)
+    YEAR_LATER = NOW + datetime.timedelta(weeks=52)
+
+    def assert_pcoll_equal(self, actual, expected):
+        """Asserts that the given PCollections are equal.
+
+        IMPORTANT: This method MUST be run under the same pipeline that produced
```

Comment:
Gosh, this is definitely going to come up. Is there any way we can prevent it -- e.g. "with pipeline" sets something when entered and assert_pcoll_equal() verifies that we're in the context of a pipeline, otherwise it errors?

Ditto for assert_pcoll_empty.

Reply:
I don't follow -- surely it's possible that some test writes the "wrong" code, below, just innocently? That's what I'm trying to guard against -- the assertion should still error in some way, in that case.

Reply:
My point is that there's no need to write the code below because `PipelinedTestBase` does the right thing already. Tests that try to set up their own pipeline simply shouldn't.

Reply:
Actually, to make this breakage-proof, I've erased `BeamTestBase` entirely and made `PipelinedTestBase` the only class available. This way, there's no public test API that allows someone to shoot themselves in the foot.